### ISSUE TITLE
NETTING flip position wrong avg_px_open

### ISSUE THREAD

**Issue by OnlyC** [nautechsystems/nautilus_trader#963]
2022-08-02T19:37:46 - I'm having a SHORT -1.1 @2.7810247713884997
2022-08-03T01:26:46 - Market BUY 375.8 @2.662. Position changed event show side=LONG, net_qty=374.7 but avg_px_open is the same @2.7810247713884997. This should be @2.662
2022-08-03T03:29:32 - Stop Market BUY 1.1 @2.781 do add to position, avg_px_open is correct calculating this time.

The **avg_px_open** looks like doesn't change when flip position.

```
**2022-08-02T19:37:46.320000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] PositionChanged(instrument_id=GTCUSDT-PERP.BINANCE, position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, account_id=BINANCE-001, opening_order_id=O-20220802-001-semiauto_ua_rt_dd_market-9, closing_order_id=None, entry=SELL, side=SHORT, net_qty=-1.1, quantity=1.1, peak_qty=367.6, currency=USDT, avg_px_open=2.7810247713884997, avg_px_close=2.7974889510489502, realized_return=-0.00592, realized_pnl=-5.44157789 USDT, unrealized_pnl=-0.00657275 USDT, ts_opened=1659459164621000000, ts_last=1659469066320000000, ts_closed=0, duration_ns=0).**
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: [CMD]--> CancelOrder(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-12, venue_order_id=BINANCE-1-012).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderInitialized(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-15, side=BUY, type=STOP_MARKET, quantity=1.1, time_in_force=GTC, post_only=False, reduce_only=True, options={'trigger_price': '2.781', 'trigger_type': 'DEFAULT', 'expire_time_ns': 0}, emulation_trigger=NO_TRIGGER, contingency_type=NO_CONTINGENCY, order_list_id=None, linked_order_ids=None, parent_order_id=None, tags=semiauto_ua_rt_dd_market,stop_close).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: [CMD]--> SubmitOrder(order=StopMarketOrder(BUY 1.1 GTCUSDT-PERP.BINANCE STOP_MARKET @ 2.781[DEFAULT] GTC, status=INITIALIZED, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-15, venue_order_id=None, tags=semiauto_ua_rt_dd_market,stop_close), position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, exec_algorithm_spec=None).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderSubmitted(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-15, account_id=BINANCE-001, ts_event=1659484336680000000).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderPendingCancel(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-12, venue_order_id=BINANCE-1-012, account_id=BINANCE-001, ts_event=1659484336680000000).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=99_974.17162303 USDT, locked=2.13631486 USDT, free=99_972.03530817 USDT)], margins=[MarginBalance(initial=2.13320681 USDT, maintenance=0.00310805 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=bd2d13d0-5c38-42c2-8003-5e712858e127).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderCanceled(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-12, venue_order_id=BINANCE-1-012, account_id=BINANCE-001, ts_event=1659484336680000000).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=99_974.17162303 USDT, locked=2.14253095 USDT, free=99_972.02909208 USDT)], margins=[MarginBalance(initial=2.13942290 USDT, maintenance=0.00310805 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=cac66b86-e915-4a99-9415-e8630a1a9a46).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderAccepted(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-15, venue_order_id=BINANCE-1-015, account_id=BINANCE-001, ts_event=1659484336680000000).
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderInitialized(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220803-001-semiauto_ua_rt_dd_market-16, side=BUY, type=MARKET, quantity=375.8, time_in_force=GTC, post_only=False, reduce_only=False, options={}, emulation_trigger=NO_TRIGGER, contingency_type=NO_CONTINGENCY, order_list_id=None, linked_order_ids=None, parent_order_id=None, tags=semiauto_ua_rt_dd_market,market_order,rt).   
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: [CMD]--> SubmitOrder(order=MarketOrder(BUY 375.8 GTCUSDT-PERP.BINANCE MARKET GTC, status=INITIALIZED, client_order_id=O-20220803-001-semiauto_ua_rt_dd_market-16, 
venue_order_id=None, tags=semiauto_ua_rt_dd_market,market_order,rt), position_id=None, exec_algorithm_spec=None).
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderSubmitted(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220803-001-semiauto_ua_rt_dd_market-16, account_id=BINANCE-001, ts_event=1659490006778000000).
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=100_018.50098028 USDT, locked=2.14253095 USDT, free=100_016.35844933 USDT)], margins=[MarginBalance(initial=2.13942290 USDT, maintenance=0.00310805 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=08706d62-a1fb-41e8-9940-3a70b7998131).
**2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderFilled(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220803-001-semiauto_ua_rt_dd_market-16, venue_order_id=BINANCE-1-016, account_id=BINANCE-001, trade_id=BINANCE-1-018, position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, order_side=BUY, order_type=MARKET, last_qty=375.8, last_px=2.662 USDT, commission=0.40015184 USDT, liquidity_side=TAKER, ts_event=1659490006778000000).**
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.Portfolio: GTCUSDT-PERP.BINANCE net_position=374.7
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=100_018.50098028 USDT, locked=3.19813625 USDT, free=100_015.30284403 USDT)], margins=[MarginBalance(initial=2.13942290 USDT, maintenance=1.05871335 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=6118892e-b37d-4c7a-92d9-e0f578a07273).
**2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] PositionChanged(instrument_id=GTCUSDT-PERP.BINANCE, position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, account_id=BINANCE-001, opening_order_id=O-20220802-001-semiauto_ua_rt_dd_market-9, closing_order_id=None, entry=BUY, side=LONG, net_qty=374.7, quantity=374.7, 
peak_qty=374.7, currency=USDT, avg_px_open=2.7810247713884997, avg_px_close=2.750810597726439, realized_return=0.01086, realized_pnl=38.88777936 USDT, unrealized_pnl=-44.59858184 USDT, ts_opened=1659459164621000000, ts_last=1659490006778000000, ts_closed=0, duration_ns=0).**
2022-08-03T03:29:32.170000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=100_018.49975664 USDT, locked=3.19192016 USDT, free=100_015.30783648 USDT)], margins=[MarginBalance(initial=2.13320681 USDT, maintenance=1.05871335 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=77e41f0f-cd3b-4bd0-a2a5-c9cbb50c3b4b).
2022-08-03T03:29:32.170000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderFilled(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-15, venue_order_id=BINANCE-1-015, account_id=BINANCE-001, trade_id=BINANCE-1-019, position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, order_side=BUY, order_type=STOP_MARKET, last_qty=1.1, last_px=2.781 USDT, commission=0.00122364 USDT, liquidity_side=TAKER, ts_event=1659497372170000000).
2022-08-03T03:29:32.170000000Z [INF] BACKTESTER-001.Portfolio: GTCUSDT-PERP.BINANCE net_position=375.8
2022-08-03T03:29:32.170000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=100_018.49975664 USDT, locked=3.19502821 USDT, free=100_015.30472843 USDT)], margins=[MarginBalance(initial=2.13320681 USDT, maintenance=1.06182140 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=4dfe9ed2-4d93-437f-8eae-ed5162505369).
**2022-08-03T03:29:32.170000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] PositionChanged(instrument_id=GTCUSDT-PERP.BINANCE, position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, account_id=BINANCE-001, opening_order_id=O-20220802-001-semiauto_ua_rt_dd_market-9, closing_order_id=None, entry=BUY, side=LONG, net_qty=375.8, quantity=375.8, 
peak_qty=375.8, currency=USDT, avg_px_open=2.781024698880444, avg_px_close=2.750810597726439, realized_return=0.01086, realized_pnl=38.88655572 USDT, unrealized_pnl=-0.00928184 USDT, ts_opened=1659459164621000000, ts_last=1659497372170000000, ts_closed=0, duration_ns=0).**

### Expected Behavior
NETTING position calculate correct avg_px_open.

### Specifications

- OS platform: Windows 64
- Python version: 3.10.5
- `nautilus_trader` version: I built from main + applied 2 fixes from https://github.com/nautechsystems/nautilus_trader/issues/933.

**Comment by OnlyC** [comment]
This led me to https://github.com/nautechsystems/nautilus_trader/issues/60 which I don't know if you have decided on the way we handle NETTING position yet !?
I go ahead and just try to modify _handle_buy_order_fill & _handle_sell_order_fill to change avg_px_open. It's working but still, when position closed, we lost realized_pnl.

**Comment by rsmb7z** [comment]
> I think this would simplify things. In my experience I didn't encounter the possibility of having multiple positions for the same instrument, on the same direction on the same broker. Most of the exchanges allow only one position per instrument. Few of them (e.g. OkEx) allow a position for LONGs and a position for SHORTs per instrument. Personally I prefer having one position per instrument per broker.

I have not experienced the above yet but just to add my thoughts on quoted comment from #60, as this is very true applicable case for NETTING and most brokers allow NETTING only - which is simpler implementation-wise as compared to HEDGING where additional orders add up in same PositionSide only, and allowing to keep both sides.
Coming to the widely used scneario of NETTING, although for simplicity we can close the previous position and then open new order which serves the flipping purpose. However one would like to to place single order, to save commissions (where not charged in percent), and 200 SHORT, to close existing 100 LONG and open new net 100 SHORT.

**Comment by cjdsellers** [comment]
This should now be fixed on `develop` from e12ae6e1743acd6c57390a8c93d9d51321c67fd9. The logic now separates the flipping fill to first close the existing position, and then open on the opposite side (just like for HEDGING).

Please let me know your findings when you test.

**Comment by OnlyC** [comment]
Yes, it's working as expected. Position closed and open a new position. realized_pnl is loss, I guess we can accumulate it in strategy.
